source/common/config/udpa_resource.cc

```diff
@@ -0,0 +1,79 @@
+#include "common/config/udpa_resource.h"
+
+#include <algorithm>
+
+#include "common/common/fmt.h"
+#include "common/http/utility.h"
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_split.h"
+
+// TODO(htuch): This file has a bunch of ad hoc URI encoding/decoding based on Envoy's HTTP util
+// functions. Once https://github.com/envoyproxy/envoy/issues/6588 lands, we can replace with GURL.
+
+namespace Envoy {
+namespace Config {
+
+using PercentEncoding = Http::Utility::PercentEncoding;
+
+std::string UdpaResourceName::encodeUri(const udpa::core::v1::ResourceName& resource_name,
+                                        const EncodeOptions& options) {
+  // We need to percent-encode authority, id, path and query params. Qualified types should not have
+  // reserved characters.
+  const std::string authority = PercentEncoding::encode(resource_name.authority(), "%/#?&=");
+  std::vector<std::string> path_components;
+  for (const auto& id_component : resource_name.id()) {
+    path_components.emplace_back(PercentEncoding::encode(id_component, "%/#?&="));
+  }
+  const std::string path = absl::StrJoin(path_components, "/");
+  std::vector<std::string> query_param_components;
+  for (const auto& context_param : resource_name.context().params()) {
+    query_param_components.emplace_back(
+        absl::StrCat(PercentEncoding::encode(context_param.first, "%#&="), "=",
+                     PercentEncoding::encode(context_param.second, "%#&=")));
+  }
+  if (options.sort_context_params_) {
+    std::sort(query_param_components.begin(), query_param_components.end());
+  }
+  const std::string query_params =
+      query_param_components.empty() ? "" : "?" + absl::StrJoin(query_param_components, "&");
+  return absl::StrCat("udpa://", authority, "/", resource_name.qualified_type(),
+                      path.empty() && query_params.empty() ? "" : "/", path, query_params);
+}
+
+udpa::core::v1::ResourceName UdpaResourceName::decodeUri(absl::string_view resource_uri) {
+  if (!absl::StartsWith(resource_uri, "udpa://")) {
```

Comment:
Not going to support the HTTP or file transports?

Reply:
This is just for resource names, which are inherently tied to the UDPA transport. Resource locators have a schema oneof which will cover these transports.

Reply:
Hmm, I didn't notice that this was for resource names as opposed to locators.  Why do we need code to convert names to URL form?  Shouldn't that only happen for locators?

Reply:
They both have a `udpa://` URI form as per the spec. I think this will be useful for debugging and possibly also for allowing later interop with control planes that don't yet understand `UdpaResourceName` extensions.